v2/apiserver/internal/core/jobs.go

```diff
@@ -253,15 +257,44 @@ func (j *jobsService) Create(
 	if err != nil {
 		return errors.Wrapf(err, "error retrieving event %q from store", eventID)
 	}
-	if _, ok := event.Worker.Job(job.Name); ok {
-		return &meta.ErrConflict{
-			Type: JobKind,
-			ID:   job.Name,
-			Reason: fmt.Sprintf(
-				"Event %q already has a job named %q.",
+	if originalJob, ok := event.Worker.Job(job.Name); ok {
+		// If we aren't dealing with a retry event, return an error
+		if event.Labels == nil || event.Labels["retryOf"] == "" {
+			return &meta.ErrConflict{
+				Type: JobKind,
+				ID:   job.Name,
+				Reason: fmt.Sprintf(
+					"Event %q already has a job named %q.",
+					eventID,
+					job.Name,
+				),
+			}
+		}
+
+		// Else, deal with retry event job
+		//
+		// If the job has already succeeded, simply return, as we don't want to
+		// replace this job in the store or schedule again
+		if originalJob.Status != nil &&
+			originalJob.Status.Phase == JobPhaseSucceeded {
+			return nil
+		}
+		// We don't currently support retrying jobs that require a shared workspace
+		if useWorkspace && event.Worker.Spec.UseWorkspace {
```

Comment:
I think the determination of whether an event can be retried needs to be made at the event-level when the retry is first requested. I have two reasons for this. The first being that it seems friendlier to inform the user up front if an event cannot be retried than to blindly attempt the retry and then have the retry fail.

The second reason is that it's not actually safe to, on a job-by-job basis, say "this job doesn't use the shared workspace so it can be retried." Here's a scenario to illustrate:

A worker is _supposed_ to run jobs A, B, and C in serial. A and C use the shared workspace. B does not.

A succeeds on the first attempt. B fails. The failed event has no record of a job named C because the worker never got as far as creating job C.

On retry, A is skipped because it succeeded. B is retried because it previously failed and _does not_ depend on the shared workspace. When B succeeds, the worker will create job C (for the first time). Job C, being dependent on the shared workspace, might have been dependent on things that job A should have done-- except job A didn't rerun, which means the shared workspace may not be in the expected state.

This is why there's a `UseWorkspace` field on the `WorkerSpec` type. We should probably just have a look at that field when retry is requested and refuse to retry if it's `true`.